test/functional/services/find.js

```diff
@@ -105,143 +81,156 @@ export function FindProvider({ getService }) {
+      log.debug(`Wait for element become visible: ${selector} with timeout=${timeout}`);
+      await driver.wait(until.elementIsVisible(element._webElement), timeout);
+      return wrap(element);
     }
 
-    async findDisplayedByLinkText(selector, timeout = defaultFindTimeout) {
-      log.debug('Find.byLinkText: ' + selector);
-      return await this._ensureElementWithTimeout(timeout, async leadfoot => {
-        return wrap(await leadfoot.findDisplayedByLinkText(selector));
-      });
+    async byLinkText(selector, timeout = defaultFindTimeout) {
+      log.debug(`Find.byLinkText('${selector}') with timeout=${timeout}`);
+      return wrap(await driver.wait(until.elementLocated(By.linkText(selector)), timeout));
     }
 
     async byPartialLinkText(partialLinkText, timeout = defaultFindTimeout) {
-      log.debug(`find.byPartialLinkText(${partialLinkText})`);
-      return await this._ensureElementWithTimeout(timeout, async leadfoot => {
-        return wrap(await leadfoot.findByPartialLinkText(partialLinkText));
-      });
+      log.debug(`Find.byPartialLinkText('${partialLinkText}')  with timeout=${timeout}`);
+      return wrap(await driver.wait(until.elementLocated(By.partialLinkText(partialLinkText)), timeout));
     }
 
     async exists(findFunction, timeout = WAIT_FOR_EXISTS_TIME) {
-      return await this._withTimeout(timeout, async leadfoot => {
-        try {
-          await findFunction(leadfoot);
-          return true;
-        } catch (error) {
-          return false;
+      await this._withTimeout(timeout);
+      try {
+        const found = await findFunction(driver);
+        await this._withTimeout(defaultFindTimeout);
+        if (Array.isArray(found)) {
+          return found.length > 0;
+        } else {
+          return found instanceof WebElementWrapper;
         }
-      });
+      } catch (err) {
+        await this._withTimeout(defaultFindTimeout);
+        return false;
+      }
     }
 
     async existsByLinkText(linkText, timeout = WAIT_FOR_EXISTS_TIME) {
-      log.debug(`existsByLinkText ${linkText}`);
-      return await this.exists(async leadfoot => wrap(await leadfoot.findDisplayedByLinkText(linkText)), timeout);
+      log.debug(`Find.existsByLinkText('${linkText}')  with timeout=${timeout}`);
+      return await this.exists(async driver => wrapAll(await driver.findElements(By.linkText(linkText))), timeout);
     }
 
     async existsByDisplayedByCssSelector(selector, timeout = WAIT_FOR_EXISTS_TIME) {
-      log.debug(`existsByDisplayedByCssSelector ${selector}`);
-      return await this.exists(async leadfoot => wrap(await leadfoot.findDisplayedByCssSelector(selector)), timeout);
+      log.debug(`Find.existsByDisplayedByCssSelector('${selector}') with timeout=${timeout}`);
+      return await this.exists(async (driver) => {
+        const elements = wrapAll(await driver.findElements(By.css(selector)));
+        return await this.filterElementIsDisplayed(elements);
```

Comment:
Filtering after the `findElements()` call like this means that the timeout isn't used if the `findElements()` call returns elements but none of them are displayed, I'm _pretty sure_ this is the reason we're seeing flakiness like https://kibana-ci.elastic.co/job/elastic+kibana+master/66/JOB=x-pack-ciGroup6,node=immutable/testReport/junit/X-Pack%20Chromium%20Functional%20Reporting%20Tests/x-pack_test_reporting_functional_reporting%C2%B7js/reporting_app_Reporting_Dashboard_Print_PDF_button_becomes_available_when_saved/

I tried to fix this like so https://github.com/elastic/kibana/pull/31368/files#diff-6238cdd3f41d65e4fddb017713976292R209, but it caused other timeout issues, since calls to `exist()` were waiting for longer, and subsequent timeout calls were missing temporary elements like the success toast here:

```
[00:20:37]                 │ debg Waiting up to 20000ms for flyout closed...
[00:20:37]                 │ debg TestSubjects.exists(dashboardAddPanel)
[00:20:37]                 │ debg Find.existsByDisplayedByCssSelector('[data-test-subj~="dashboardAddPanel"]') with timeout=2500
[00:20:48]                 │ debg Find.findByCssSelector('.euiToast') with timeout=10000
...
[00:20:58]                 └- ✖ fail: "dashboard app using current data dashboard snapshots compare area chart snapshot"
[00:20:58]                 │      TimeoutError: Waiting for element to be located By(css selector, .euiToast)
[00:20:58]                 │ Wait timed out after 10033ms
[00:20:58]                 │       at node_modules/selenium-webdriver/lib/webdriver.js:834:17
[00:20:58]                 │       at process._tickCallback (internal/process/next_tick.js:68:7)
```
https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1118/JOB=kibana-ciGroup3,node=immutable/testReport/junit/UI%20Functional%20Tests/test_functional_apps_dashboard__dashboard_snapshots%C2%B7js/dashboard_app_using_current_data_dashboard_snapshots_compare_area_chart_snapshot/